src/generate/rust.rs

```diff
@@ -1492,3 +813,540 @@ impl<Pat: Ord + Hash + RustInputPat> Rule<Pat> {
         }
     }
 }
+
+fn impl_parse_with(name: &str, rust_slice_ty: &str, out: &mut String) {
+    put!(out, "
+    impl<'a, 'i, I: ::gll::runtime::Input<Slice = ", rust_slice_ty, ">> ", name, "<'a, 'i, I> {
+        pub fn parse_with<R>(
+            input: I,
+            f: impl for<'b, 'i2> FnOnce(
+                &'b ::gll::runtime::Parser<'i2, _P, _C, I>,
+                ParseResult<'b, 'i2, I, ", name, "<'b, 'i2, I>>,
+            ) -> R,
+        ) -> R {
+            ::gll::runtime::Parser::with(input, |mut parser, range| {
+                let call = Call {
+                    callee: ", CodeLabel(name.to_string()), ",
+                    range,
+                };
+                parser.threads.spawn(
+                    Continuation {
+                        code: call.callee,
+                        fn_input: call.range,
+                        state: 0,
+                    },
+                    call.range,
+                );
+                parse(&mut parser);
+                let result = parser.memoizer.longest_result(call);
+                f(&parser, result.ok_or(ParseError::NoParse).and_then(|range| {
+                    let handle = Handle {
+                        node: ParseNode { kind: ", ParseNodeKind(name.to_string()), ", range },
+                        parser: &parser,
+                        _marker: PhantomData,
+                    };
+                    if range == call.range {
+                        Ok(handle)
+                    } else {
+                        Err(ParseError::TooShort(handle))
+                    }
+                }))
+            })
+        }
+    }")
+}
+
+fn declare_rule<Pat>(
+    name: &str,
+    rule: &RuleWithNamedFields<Pat>,
+    parse_nodes: &RefCell<ParseNodeMap<Pat>>,
+    out: &mut String,
+)
+where
+    Pat: Ord + Hash + RustInputPat,
+{
+    let variants = rule.find_variant_fields();
+
+    // Declare rule type
+    if let Some(variants) = &variants {
+        put!(out, "pub enum ", name, "<'a, 'i: 'a, I: 'a + ::gll::runtime::Input> {");
+        for Variant { rule, name: variant_name, fields } in variants {
+            if fields.is_empty() {
+                put!(out, variant_name, "(Handle<'a, 'i, I, ", rule.field_type(&[]), ">),");
+            } else {
+                put!(out, variant_name, " {");
+                for (field_name, paths) in fields {
+                    let refutable = rule.field_pathset_is_refutable(paths);
+                    put!(out, field_name, ": ");
+                    if refutable {
+                        put!(out, "Option<");
+                    }
+                    put!(out, "Handle<'a, 'i, I, ", rule.field_pathset_type(paths), ">");
+                    if refutable {
+                        put!(out, ">");
+                    }
+                    put!(out, ",");
+                }
+                put!(out, "},");
```

Comment:
`out` is repeated a lot, this is why I really liked the macro without it. In fact, I'd rather call it `p!` now or something short like that.